File: IAIso-v5.0/core/iaiso/conformance/_events.py
"""Events vector runner.

For each vector, construct an engine, feed in the steps, capture emitted
events, and compare against the expected stream per spec/events/README.md §5.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from iaiso.conformance import ScriptedClock, VectorResult, load_vectors


def run_events_vectors(spec_root: Path) -> list[VectorResult]:
    from iaiso.audit import MemorySink
    from iaiso.core.engine import PressureConfig, PressureEngine, StepInput

    data = load_vectors(spec_root, "events")
    tolerance = float(data.get("tolerance", 1e-9))
    results: list[VectorResult] = []

    for vec in data["vectors"]:
        name = vec["name"]
        try:
            results.append(_run_events_vector(
                name, vec, tolerance, PressureConfig, PressureEngine, StepInput, MemorySink,
            ))
        except Exception as exc:  # noqa: BLE001
            results.append(VectorResult(
                section="events", name=name, passed=False,
                message=f"runner exception: {type(exc).__name__}: {exc}",
            ))

    return results


def _run_events_vector(
    name, vec, tolerance, PressureConfig, PressureEngine, StepInput, MemorySink,
):
    config = PressureConfig(**vec.get("config", {}))
    clock = ScriptedClock(vec["clock"])
    audit = MemorySink()
    execution_id = vec.get("execution_id", f"exec-events-{name}")

    engine = PressureEngine(
        config,
        execution_id=execution_id,
        audit_sink=audit,
        clock=clock,
    )

    reset_after = vec.get("reset_after_step")
    for i, step in enumerate(vec.get("steps", [])):
        engine.step(StepInput(
            tokens=step.get("tokens", 0),
            tool_calls=step.get("tool_calls", 0),
            depth=step.get("depth", 0),
            tag=step.get("tag"),
        ))
        if reset_after == i + 1:
            new_clock = vec.get("clock_after_reset")
            if new_clock is not None:
                clock._values.insert(clock._index, new_clock)
            engine.reset()

    emitted = [_event_to_dict(e) for e in audit.events]

    # Negative assertions: some vectors say the stream MUST NOT contain
    # a kind that, given the config, shouldn't appear. We check by
    # seeing if the expected_events list is a prefix or full match.
    expected = vec["expected_events"]

    # Check length: emitted must contain AT LEAST len(expected) events
    if len(emitted) < len(expected):
        return VectorResult(
            section="events", name=name, passed=False,
            message=f"expected at least {len(expected)} events, got {len(emitted)}",
            details={"emitted": emitted, "expected": expected},
        )

    # For a strict spec, a conformant impl should emit EXACTLY the
    # expected events — not extras, not fewer — UNLESS the vector says
    # otherwise. We use strict equality unless `expected_at_least` is set.
    strict = vec.get("strict_length", True)
    if strict and len(emitted) != len(expected):
        return VectorResult(
            section="events", name=name, passed=False,
            message=f"expected exactly {len(expected)} events, got {len(emitted)}",
            details={"emitted": emitted, "expected": expected},
        )

    for i, (e_actual, e_expected) in enumerate(zip(emitted, expected)):
        mismatch = _compare_event(e_actual, e_expected, tolerance)
        if mismatch:
            return VectorResult(
                section="events", name=name, passed=False,
                message=f"event[{i}] ({e_expected.get('kind', '?')}): {mismatch}",
                details={"actual": e_actual, "expected": e_expected},
            )

    return VectorResult(section="events", name=name, passed=True)


def _event_to_dict(event) -> dict[str, Any]:
    return {
        "schema_version": event.schema_version,
        "execution_id": event.execution_id,
        "kind": event.kind,
        "timestamp": event.timestamp,
        "data": dict(event.data),
    }


def _compare_event(actual: dict, expected: dict, tolerance: float) -> str | None:
    """Return None if matches, else a mismatch message. See spec §5: compare
    schema_version, execution_id, kind, and every payload field listed in
    `expected`. Ignore `timestamp`. Ignore unspecified payload fields."""

    for key in ("schema_version", "execution_id", "kind"):
        if key not in expected:
            continue
        if actual.get(key) != expected[key]:
            return f"{key}: expected {expected[key]!r}, got {actual.get(key)!r}"

    expected_data = expected.get("data", {})
    actual_data = actual.get("data", {})
    for k, v in expected_data.items():
        if k not in actual_data:
            return f"data missing required key {k!r}"
        av = actual_data[k]
        if isinstance(v, (int, float)) and isinstance(av, (int, float)):
            if abs(float(av) - float(v)) > tolerance:
                return f"data.{k}: expected {v!r}, got {av!r}"
        elif av != v:
            return f"data.{k}: expected {v!r}, got {av!r}"

    return None

File: IAIso-v5.0/core/iaiso-node/src/conformance/events.ts
/**
 * Conformance runner — events vectors.
 *
 * Validates emitted event streams against the expectations in
 * spec/events/vectors.json. Compares schema_version, execution_id, kind,
 * and the payload fields explicitly listed in each expected event;
 * ignores timestamps and extra implementation-specific fields.
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";

import { AuditEvent } from "../audit/event.js";
import { MemorySink } from "../audit/sinks/memory.js";
import { PressureConfig, PressureEngine, StepInput } from "../core/engine.js";
import type { Clock } from "../core/types.js";
import type { VectorResult } from "./pressure.js";

interface EventsVector {
  name: string;
  description?: string;
  config?: Record<string, unknown>;
  clock?: number[];
  steps?: Array<{
    tokens?: number;
    tool_calls?: number;
    depth?: number;
    tag?: string | null;
  }>;
  reset_after_step?: number;
  clock_after_reset?: number;
  execution_id?: string;
  expected_events: Array<{
    schema_version?: string;
    execution_id?: string;
    kind?: string;
    data?: Record<string, unknown>;
  }>;
  strict_length?: boolean;
}

interface EventsVectorFile {
  version: string;
  tolerance?: number;
  vectors: EventsVector[];
}

export function runEventsVectors(specRoot: string): VectorResult[] {
  const data = JSON.parse(
    readFileSync(join(specRoot, "events", "vectors.json"), "utf8"),
  ) as EventsVectorFile;
  const tolerance = data.tolerance ?? 1e-9;
  const results: VectorResult[] = [];

  for (const vec of data.vectors) {
    try {
      results.push(runOneVector(vec, tolerance));
    } catch (exc) {
      results.push({
        section: "events",
        name: vec.name,
        passed: false,
        message: `runner exception: ${(exc as Error).name}: ${(exc as Error).message}`,
      });
    }
  }

  return results;
}

function runOneVector(vec: EventsVector, tolerance: number): VectorResult {
  const config = new PressureConfig(vec.config ?? {});
  const scripted = [...(vec.clock ?? [0.0])];
  let ci = 0;
  const clock: Clock = () => {
    if (ci >= scripted.length) throw new Error("clock exhausted");
    return scripted[ci++]!;
  };
  const audit = new MemorySink();
  const executionId = vec.execution_id ?? `exec-events-${vec.name}`;

  const engine = new PressureEngine(config, {
    execution_id: executionId,
    audit_sink: audit,
    clock,
  });

  const steps = vec.steps ?? [];
  for (let i = 0; i < steps.length; i++) {
    const s = steps[i]!;
    engine.step(
      new StepInput({
        tokens: s.tokens ?? 0,
        tool_calls: s.tool_calls ?? 0,
        depth: s.depth ?? 0,
        tag: s.tag ?? null,
      }),
    );
    if (vec.reset_after_step === i + 1) {
      if (vec.clock_after_reset !== undefined) {
        scripted.splice(ci, 0, vec.clock_after_reset);
      }
      engine.reset();
    }
  }

  const emitted = audit.events;
  const expected = vec.expected_events;
  const strict = vec.strict_length ?? true;

  if (emitted.length < expected.length) {
    return {
      section: "events",
      name: vec.name,
      passed: false,
      message: `expected at least ${expected.length} events, got ${emitted.length}`,
    };
  }
  if (strict && emitted.length !== expected.length) {
    return {
      section: "events",
      name: vec.name,
      passed: false,
      message: `expected exactly ${expected.length} events, got ${emitted.length}`,
    };
  }

  for (let i = 0; i < expected.length; i++) {
    const a = emitted[i]!;
    const e = expected[i]!;
    const mismatch = compareEvent(a, e, tolerance);
    if (mismatch) {
      return {
        section: "events",
        name: vec.name,
        passed: false,
        message: `event[${i}] (${e.kind ?? "?"}): ${mismatch}`,
      };
    }
  }

  return { section: "events", name: vec.name, passed: true, message: "" };
}

function compareEvent(
  actual: AuditEvent,
  expected: {
    schema_version?: string;
    execution_id?: string;
    kind?: string;
    data?: Record<string, unknown>;
  },
  tolerance: number,
): string | null {
  if (expected.schema_version !== undefined && actual.schemaVersion !== expected.schema_version) {
    return `schema_version: expected ${JSON.stringify(expected.schema_version)}, got ${JSON.stringify(actual.schemaVersion)}`;
  }
  if (expected.execution_id !== undefined && actual.executionId !== expected.execution_id) {
    return `execution_id: expected ${JSON.stringify(expected.execution_id)}, got ${JSON.stringify(actual.executionId)}`;
  }
  if (expected.kind !== undefined && actual.kind !== expected.kind) {
    return `kind: expected ${JSON.stringify(expected.kind)}, got ${JSON.stringify(actual.kind)}`;
  }

  const expectedData = expected.data ?? {};
  for (const [k, v] of Object.entries(expectedData)) {
    if (!(k in actual.data)) {
      return `data missing required key ${JSON.stringify(k)}`;
    }
    const av = actual.data[k];
    if (typeof v === "number" && typeof av === "number") {
      if (Math.abs(av - v) > tolerance) {
        return `data.${k}: expected ${JSON.stringify(v)}, got ${JSON.stringify(av)}`;
      }
    } else if (JSON.stringify(av) !== JSON.stringify(v)) {
      return `data.${k}: expected ${JSON.stringify(v)}, got ${JSON.stringify(av)}`;
    }
  }

  return null;
}
